src/formatter.rs

```diff
@@ -5,60 +5,66 @@ use std::io::BufRead;
 use ansi_term::{ANSIString, ANSIStrings};
 use log::debug;
 
-use crate::config::Config;
+use crate::config::{Config, StyleConfig};
 use crate::tokenizer::Tokenizer;
 use crate::types::LineType;
 
-fn highlight_command<'a>(
-    command: &'a str,
-    example_code: &'a str,
-    config: &Config,
+fn highlight_code_segment<'a>(
+    command_name: &'a str,
+    segment: &'a str,
+    name_painter: &impl Fn(&'a str) -> ANSIString<'a>,
+    code_painter: &impl Fn(&'a str) -> ANSIString<'a>,
     parts: &mut Vec<ANSIString<'a>>,
 ) {
-    let mut code_part_end_pos = 0;
-    while let Some(command_start) = example_code[code_part_end_pos..].find(&command) {
-        let code_part = &example_code[code_part_end_pos..code_part_end_pos + command_start];
-        parts.push(config.style.example_code.paint(code_part));
-        if code_part_end_pos == 0 {
-            // Only highlight command names at the start of the line ...
-            parts.push(config.style.command_name.paint(command));
+    let split_into_normal_and_command = |part: &'a str| -> (&'a str, &'a str) {
+        // Splits a segment that may end with the command name into normal text and that command
+        // name. The command name should not be separated if it appears within normal text by
+        // coincidence, thus we check that it has some leading whitespace.
+        if !part.ends_with(command_name) {
+            (part, "")
+        } else if part.len() == command_name.len() {
+            ("", part)
         } else {
-            let char_before_command = example_code
-                .chars()
-                .nth(code_part_end_pos + command_start - 1);
-            if char_before_command.filter(|c| c.is_whitespace()).is_some() {
-                // ... or when preceded by a whitespace character.
-                parts.push(config.style.command_name.paint(command));
-            } else {
-                parts.push(config.style.example_code.paint(command));
-            }
+            let split_point = part
+                .char_indices()
+                .nth(part.chars().count() - command_name.chars().count() - 1)
```

Comment:
I see. We should create a test for non-ASCII input.

Reply:
Fortunately Rust makes it quite hard to mis-use unicode. For example, strings do not allow indexing (to avoid the chance of indexing on a non-boundary) 🙂 So in practice, as long as you're not unwrapping fallible functions, I've never seen a Rust codebase with any of the "simple" unicode bugs.